examples/permissioned-token/assets/permissions.py

```diff
@@ -0,0 +1,129 @@
+        a) receiver asset balance should be <=100 after transfer
+        b) both [from, to] accounts must be whitelisted
+    Permissions manager manages the contract permission.
+    It can be changed by the asset manager
+    """
+
+    true = Int(1)
+    max_tokens = Bytes("max_tokens") # maximum token transfer rule
+    whitelist_count = Bytes("whitelist_count") # whitelisted accounts counter
+    controller_app_id = Bytes("controller_app_id") # token's controller application index
+
+    on_deployment = Seq([
+	Assert(And(
+		Txn.application_args.length() == Int(1),
+
+		# Always verify that the RekeyTo property of any transaction is set to the ZeroAddress
+		# unless the contract is specifically involved ina rekeying operation.
+		Txn.rekey_to() == Global.zero_address()
+        )),
+
+        # Save max_tokens count in global state (default = 100, during ssc deploy)
+        App.globalPut(max_tokens, Int(100)),
+
+        # Initialize whitelisted accounts counter to 0
+        App.globalPut(whitelist_count, Int(0)),
+
+        # Save controller application index in global state
+        App.globalPut(controller_app_id, Btoi(Txn.application_args[0])),
+        Return(Int(1))
+    ])
+
+    # fetch permissions manager, token_id from controller's global state (using foreignApps)
+    permission_manager = App.globalGetEx(Int(1), Bytes("permissions_manager"))
+    token_id = App.globalGetEx(Int(1), Bytes("token_id"))
+
+    # add an account to be whitelisted (token transfer is rejected if account is not whitelisted)
+    add_whitelist = Seq([
+        permission_manager,
+        token_id,
+        Assert(And(
+		Txn.application_args.length() == Int(2),
+		Txn.accounts.length() == Int(1),
+		Txn.rekey_to() == Global.zero_address(),
+		#Txn.applications.length() ==  Int(1), [TEALv3]
+
+		# verify from global state whether controller application passed is the valid one
+		# note: with tealv3 we can just use txn.applications[0] instead of passing an an appArg
+		Btoi(Txn.application_args[1]) == App.globalGet(controller_app_id),
+
+		# first verify correct token value is passed
+		# token_id.hasValue(),
+			# Txn.assets[0] == token_id.value(), [TEALv3]
+
+		# then verify txn sender is the token manager
+		permission_manager.hasValue(),
+		Txn.sender() == permission_manager.value(),
+        )),
+
+        If(
+            # If account is not already whitelisted, update the counter and continue, return otherwise
+            App.localGet(Int(1), Bytes("whitelisted")) == Int(0),
+            App.globalPut(whitelist_count, App.globalGet(whitelist_count) + Int(1)), # only evaluated if above is true
+            Return(Int(1)) # else return
+        ),
+
+        # finally update txn.accounts[1] local state to set whitelisted status as true(1)
+        App.localPut(Int(1), Bytes("whitelisted"), Int(1)),
+        Return(Int(1))
+    ])
+
+    asset_balance = AssetHolding.balance(Int(1), Gtxn[1].xfer_asset())
+    transfer_token = Seq([
```

Comment:
Can you explain why we are checking for this case in every contract?

Reply:
This is the most important check. Please read the README file with the mechanism description.